### ISSUE TITLE
The nccl P2P test reflects no bw & latency differences with or without GDR

### ISSUE THREAD

**Issue by Moson864** [NVIDIA/nccl#1188]
We are testing the performance of SendRecv in nccl-tests with 2 servers, 8 GPUS per server.
We enable GDR through modprobe nvidia-peermem and enable the gdrcopy at the same time, shown as the following fig:
![nvidia-peer](https://github.com/NVIDIA/nccl/assets/84839424/16a63b5c-ee43-4409-ac97-59ba8081d26e)
However, it appears that even we enable the gdr and the printed log seems already enabling gdr, the result busbw and time are not really different from cases without gdr:
**without gdr:**
![NGDR-nccltests](https://github.com/NVIDIA/nccl/assets/84839424/833c12bc-8043-49cf-b931-6aafbfe19afb)
**with gdr:**
![GDR-nccltests](https://github.com/NVIDIA/nccl/assets/84839424/ccaf1768-a61d-49f2-93ae-fd6a96afb359)
We have already employed simple commands without other additional environment variables:
`mpirun -np 2 \
	-H node1:1,node2:1\
	--allow-run-as-root -bind-to none -map-by slot \
	-x LD_LIBRARY_PATH -x PATH \
	-x NCCL_DEBUG=INFO -x NCCL_IB_DISABLE=0 -x NCCL_SOCKET_IFNAME=ens2f0 \
	-x NCCL_IB_HCA=mlx5_0:1 \
	./sendrecv_perf -b 1 -e 4M -f 2`
We choose the GPU0 on both sides and you could clearly see from the topo below that the path between GPU & NIC is PATH_PIX on both sides, still the gdr doesn't show any imporvement.
It really confuses us because we have tested on osu-benchmarks and observed about 4x bandwidth improvement on the same environment. Commands and Results are as follows:
`mpirun -np 2 -H node1:1,node2:1 --allow-run-as-root -bind-to none -map-by slot \
	-x UCX_NET_DEVICES=mlx5_0:1  \
	-x UCX_RNDV_SCHEME=get_zcopy \
	-x UCX_IB_GPU_DIRECT_RDMA=1 \
    	-x LD_LIBRARY_PATH \
	-x UCX_RNDV_THRESH=128 \
	-x UCX_TLS=rc,cuda_copy,gdr_copy \
	./osu_bw -d cuda D D
`
**without gdr:**
![NGDR-osu](https://github.com/NVIDIA/nccl/assets/84839424/c73ecee4-efa4-4dde-b6da-c3bbdd864d5d)
**without gdr:**
![GDR-osu](https://github.com/NVIDIA/nccl/assets/84839424/09b996a2-7423-4045-acfc-b61f5197fe57)

Also we have tried many approaches:
1. set NCCL_GDRCOPY_ENABLE & NCCL_GDRCOPY_FLUSH_ENABLE;
2. use ucx plugin & set the RNDV_THRESH as osu;
3. set NCCL_NET_GDR_READ=1;
4. change the NCCL_PROTO;
Still, we didn't get any differences with or without gdr.
We are not sure what's wrong with our tests, looking forward to your replies. Many thanks. @sjeaugey 
**Environment:**
**Hardware:**
GPU: AMD RX6500
HCA: Mellanox Technologies MT27710   [ConnectX-4 Lx]
**Software**
cuda11.7
MLNX_OFED_LINUX-5.5-1.0.3.2
openmpi-4.1.2
nccl_2.19.4
gdrcopy v2.4.1
**Topo**
![topo](https://github.com/NVIDIA/nccl/assets/84839424/35016c69-9f38-45ea-a509-ba3918ef415f)

**Comment by sjeaugey** [comment]
What is your network port speed? Is it 25Gbps? If so, the peak is 3GB/s and NCCL achieves 2.7GB/s. You may want to increase the max size to `1G`, to see the real peak of NCCL (seems numbers are still increasing), but to me, it looks like numbers are what they should be.

The fact NCCL can achieve the same performance without GDR is not that surprising, given the network speed is much lower than the PCI speed, so we can afford a back and forth to the CPU, and the extra CPU memory bandwidth.

I'm not sure what PCI bandwidth you have though, given you mentioned your GPU was an "AMD RX6500", but it should be at least PCI Gen3x16 which would be 12GB/s -- more than enough to cover an extra round trip of a 3GB/s flow.

**Comment by Moson864** [comment]
> What is your network port speed? Is it 25Gbps? If so, the peak is 3GB/s and NCCL achieves 2.7GB/s. You may want to increase the max size to `1G`, to see the real peak of NCCL (seems numbers are still increasing), but to me, it looks like numbers are what they should be.
> 
> The fact NCCL can achieve the same performance without GDR is not that surprising, given the network speed is much lower than the PCI speed, so we can afford a back and forth to the CPU, and the extra CPU memory bandwidth.
> 
> I'm not sure what PCI bandwidth you have though, given you mentioned your GPU was an "AMD RX6500", but it should be at least PCI Gen3x16 which would be 12GB/s -- more than enough to cover an extra round trip of a 3GB/s flow.

Thanks for your replies!!!
Our network port speed is 25Gbps and PCIe Gen4x16 speed is 16GT/s. Absolutely the maximun bandwidth could achieve the theoretical value with or without gdr.
However, compared with osu results, the bandwidth result of nccl-tests of the middle data segment doesn't show any differences with or without gdr. As we can see, for 8192B data size, osu results reflect obvious better bandwidth performance with gdr than withour gdr. But the nccl-tests results still seem quite similar.
I wonder if there could be any approaches i could use to reveal the function of gdr in nccl-tests. Maybe use larger data size?
Many Thanks! @sjeaugey

**Comment by sjeaugey** [comment]
Ok so PCI gen4 can reach 24GB/s; 3GB/s is very small compared to that.

If you wanted to mimic the OSU test with NCCL, you'd want to run the NCCL perf test with `-R 1` so that buffers are registered on the NIC and we send them directly. Otherwise NCCL will first copy the buffer to an intermediate buffer, then send through the NIC. That has the advantage that as you can see, peak BW is good even when we stage through the CPU instead of through the CPU, but also adds an extra step which can make performance lower than OSU test for medium sizes.

**Comment by Moson864** [comment]
Thanks for your replies!!! @sjeaugey 
I'v probably understood you that for the intermediate proxy trhead in NCCL alg, one more step is needed for communicating connection and send/recv operations.
Do you mean running the sendrecv_perf with `-R 1` option? I just have a check that there might not be this option in nccl-tests. I was hoping if you could provide an example.  Sry for disturbing u.

**Comment by sjeaugey** [comment]
Ah, right we haven't added that option to the NCCL perf tests yet. My bad.

**Comment by Moson864** [comment]
> Ah, right we haven't added that option to the NCCL perf tests yet. My bad.

I guess there might be some environment variables to disable the intermediate buffer so that we could use send/recv buffer directly to register RDMA mr. Have these variables already been committed into the newest version of nccl? Maybe I could change them and regenerate the lib.
Besides, i wonder if there could be some earlier versions of nccl that registers send/recv buffer directly and supports GDR at the same time?
Many thanks!!! @sjeaugey

**Comment by sjeaugey** [comment]
No, there is no such environment variables. The possibility to use user buffers directly is a new feature in NCCL 2.19 and requires the use of specific new APIs.

**Comment by Moson864** [comment]
Yes, I've found the feature ncclMemAlloc in NCCL 2.19 doc. I notice that user buffer registration is supported for NVLS. Here are my questions:
1. Does it mean that my environment must support NVLink and SHARP feature(as i find from NCCL SHARP plugin page, requirements for Mellanox ConnectX 6 HCA ) to enable this user-buffer-registration feature?
2. What's the difference between user buffer registration and so-called local based buffer registration?
Thanks for your replies! Wish you great weekend! @sjeaugey

**Comment by sjeaugey** [comment]
You don't need to use ncclMemAlloc for direct send/recv on the network, only to call ncclCommRegister() on the send/recv buffers before calling send/recv.

**Comment by Moson864** [comment]
I tried ncclMemAlloc and still didn't get any improvements for medium sizes.
I suppose that is because my environment does not support NVLink and SHARP now. I could plug in nccl-SHARP, but the INFO LOG shows that CollNet is not supported:
![SAHRP-support](https://github.com/NVIDIA/nccl/assets/84839424/c323a86e-4e4a-449c-be12-a4b6f59412d2)
Could that be the main reason for my nccl-tests results of medium sizes?
Really appreciate for your replies!!! @sjeaugey

**Comment by Moson864** [comment]
BTW，from our recent test results, we have been thinking the effects of the different hardware topo connections between NIC & GPU on GDR performance. As we could see from Nvidia GDR doc, PATH_PIX would perform best, then PATH_PHB, PATH_SYS last. Considering PCIe host bridge is a "better" PCIe switch, if we do not occupy too many CPU bus resources, could PATH_PHB reaches the same performance as PATH_PIX? Besides, if we connect GPU and NIC directly to CPU PCIe slots without normal PCIe switches between them, could that make better performance?
Many thanks for your replies! @sjeaugey
